# E-Bay Used Car Sales Data Analysis  #

Car market pricing has fluctuated heavily recently due to current market situations. Overall, used cars websites tend to offer the best value-to-price ratios for those who wish to purchase cars. 

The main objective of this research is to get a better understanding of what type of car should be purchased for prospective buyers looking to score a cheap car in good condition.

## Context ## 

The data was originally scraped in 2016 by user [orgesleka](https://www.kaggle.com/orgesleka) on [Kaggle](https://www.kaggle.com/) to generate a collection consisting of over 370,000 used car data from eBay Kleinanzeigen, a classifieds section of the German eBay website. The dataset is available [here](https://data.world/data-society/used-cars-data). Since this data was scraped from a German site, the contents of this file are in German. 

## Investigation ## 

### Data Description ###
Our dataset consists of the following columns. The descriptions of these columns are provided for reference below: 

| Header | Description |
| --- | --- |
| dateCrawled | When the ad was first crawled. All field values are taken from this date. | 
| name | Name of the car |
| seller | Whether the seller is private or a dealer | 
| offerType | Type of listing | 
| price | Price on the ad to sell the car for | 
| abtest | Whether the listing is included in an A/B test | 
| vehicleType | Vehicle type | 
| yearOfRegistration | Year in which the car was first registered | 
| gearbox | Transmission type | 
| powerPS | Power of the car in PS | 
| model | Car model name | 
| kilometer | How many km the car has driven | 
| monthOfRegistration | Month in which car was first registered |
| fuelType | What type of fuel car uses | 
| brand | Brand of the car |
| notRepairedDamage | If the car has a damage not yet repaired | 
| dateCreated | Date of eBay listing creation | 
| nrOfPictures | The number of pictures in the ad | 
| postalCode | Postal code for the location of the vehicle | 
| lastSeenOnline | When the crawler last saw the ad online | 

To begin, we will import our libraries `pandas` and `numpy` for our investigation. We read the file in from `autos.csv` with encoding `Windows-1252` to create a DataFrame object.

In [1]:
import pandas
import numpy

autos = pandas.read_csv('autos.csv', encoding="Windows-1252")

Before processing the data, to get a better understanding of the structure of our new DataFrame `autos`, we can run the inspect the dataframe information to see how many values of each column we get back and what type of data is in each column.

In [2]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

We can also take a peek at the first few rows of the DataFrame to confirm what data we expect to process.

In [3]:
autos.head()

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

From this, we can see that the dataset comes with 371,528 rows of used car data and consists of 20 columns, most of which are of type `String`, with some additional columns of type `int`. Additionally, some rows have null values for certain column data.

### Data Cleansing ###

#### Discarding Columns ####

From the data provided above, one of the first things we notice is that the column names are given in camelcase rather than Python's preferred snakecase. We can clean up the column names by converting the format and making certain column names more straightforward.

In [4]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [5]:
new_columns = []
for column in autos.columns:
    if column == "yearOfRegistration":
        new_columns.append("registration_year")
    elif column == "monthOfRegistration":
        new_columns.append("registration_month")
    elif column == "notRepairedDamage":
        new_columns.append("unrepaired_damage")
    elif column == "dateCreated":
        new_columns.append("ad_created")
    elif column == "kilometer":
        new_columns.append("odometer_km")
    else:
        newName = ""
        for char in column:
            if char.isupper():
                newName += "_" + char.lower()
            else:
                newName += char
        new_columns.append(newName)
autos.columns = new_columns
autos.head()

date_crawled                            name  seller offer_type  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat    Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat    Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat    Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat    Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat    Angebot   

   price abtest vehicle_type  registration_year    gearbox  power_p_s  model  \
0    480   test          NaN               1993    manuell          0   golf   
1  18300   test        coupe               2011    manuell        190    NaN   
2   9800   test          suv               2004  automatik        163  grand   
3   1500   test   kleinwagen               2001    manuell         75   golf   
4   3600   test   kleinwagen               2008    manuell         69  fabia   

   odometer_km  registration_month fuel_type       brand unrepaired_damage  \
0       150000                   0    benzin  volkswagen               NaN   
1       125000                   5    diesel        audi                ja   
2       125000                   8    diesel        jeep               NaN   
3       150000                   6    benzin  volkswagen              nein   
4        90000                   7    diesel       skoda              nein   

            ad_created  nr_of_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00               0        70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00               0        66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00               0        90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00               0        91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00               0        60437  2016-04-06 10:17:21

With our column names cleaned up, we can now inspect the data information of each column for data cleansing. 

In [6]:
autos.describe(include="all")

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox      power_p_s  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model    odometer_km  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

From the information above, we can see that the columns `seller`, `offer_type`, and `nr_of_pictures` seem to have mostlly only one value, and can be removed from consideration for our analysis purposes. Additionally, we do not have context regarding the tests run for the column `abtest` column, so it is not useful for our purposes either. Let's take a closer look at these columns below:

In [7]:
autos["seller"].value_counts()

privat        371525
gewerblich         3
Name: seller, dtype: int64

In [8]:
autos["offer_type"].value_counts()

Angebot    371516
Gesuch         12
Name: offer_type, dtype: int64

In [9]:
autos["nr_of_pictures"].value_counts()

0    371528
Name: nr_of_pictures, dtype: int64

In [10]:
autos["abtest"].value_counts()

test       192585
control    178943
Name: abtest, dtype: int64

Since we have only 3 listings in the `seller` column from dealerships(`gewerblich`) to compare with the large number of private(`privat`) listings, this will not give us useful analysis data. The same logic applies to `offer_type` and `nr_of_pictures`. Additionally, we can remove `abtest` as we do not have insight into the differences between the `test` and `control` groups. 

In [11]:
autos.drop(["seller", "offer_type", "nr_of_pictures", "abtest"], axis=1, inplace=True)
autos

date_crawled                                          name  \
0       2016-03-24 11:52:17                                    Golf_3_1.6   
1       2016-03-24 10:58:45                          A5_Sportback_2.7_Tdi   
2       2016-03-14 12:52:21                Jeep_Grand_Cherokee_"Overland"   
3       2016-03-17 16:54:04                            GOLF_4_1_4__3TÜRER   
4       2016-03-31 17:25:20                Skoda_Fabia_1.4_TDI_PD_Classic   
...                     ...                                           ...   
371523  2016-03-14 17:48:27                    Suche_t4___vito_ab_6_sitze   
371524  2016-03-05 19:56:21         Smart_smart_leistungssteigerung_100ps   
371525  2016-03-19 18:57:12            Volkswagen_Multivan_T4_TDI_7DC_UY2   
371526  2016-03-20 19:41:08                        VW_Golf_Kombi_1_9l_TDI   
371527  2016-03-07 19:39:19  BMW_M135i_vollausgestattet_NP_52.720____Euro   

        price vehicle_type  registration_year    gearbox  power_p_s  \
0         480          NaN               1993    manuell          0   
1       18300        coupe               2011    manuell        190   
2        9800          suv               2004  automatik        163   
3        1500   kleinwagen               2001    manuell         75   
4        3600   kleinwagen               2008    manuell         69   
...       ...          ...                ...        ...        ...   
371523   2200          NaN               2005        NaN          0   
371524   1199       cabrio               2000  automatik        101   
371525   9200          bus               1996    manuell        102   
371526   3400        kombi               2002    manuell        100   
371527  28990    limousine               2013    manuell        320   

              model  odometer_km  registration_month fuel_type  \
0              golf       150000                   0    benzin   
1               NaN       125000                   5    diesel   
2             grand       125000                   8    diesel   
3              golf       150000                   6    benzin   
4             fabia        90000                   7    diesel   
...             ...          ...                 ...       ...   
371523          NaN        20000                   1       NaN   
371524       fortwo       125000                   3    benzin   
371525  transporter       150000                   3    diesel   
371526         golf       150000                   6    diesel   
371527      m_reihe        50000                   8    benzin   

                 brand unrepaired_damage           ad_created  postal_code  \
0           volkswagen               NaN  2016-03-24 00:00:00        70435   
1                 audi                ja  2016-03-24 00:00:00        66954   
2                 jeep               NaN  2016-03-14 00:00:00        90480   
3           volkswagen              nein  2016-03-17 00:00:00        91074   
4                skoda              nein  2016-03-31 00:00:00        60437   
...                ...               ...                  ...          ...   
371523  sonstige_autos               NaN  2016-03-14 00:00:00        39576   
371524           smart              nein  2016-03-05 00:00:00        26135   
371525      volkswagen              nein  2016-03-19 00:00:00        87439   
371526      volkswagen               NaN  2016-03-20 00:00:00        40764   
371527             bmw              nein  2016-03-07 00:00:00        73326   

                  last_seen  
0       2016-04-07 03:16:57  
1       2016-04-07 01:46:50  
2       2016-04-05 12:47:46  
3       2016-03-17 17:40:17  
4       2016-04-06 10:17:21  
...                     ...  
371523  2016-04-06 00:46:52  
371524  2016-03-11 18:17:12  
371525  2016-04-07 07:15:26  
371526  2016-03-24 12:45:21  
371527  2016-03-22 03:17:10  

[371528 rows x 16 columns]

#### Translations ####
Since the original dataset is in German, we should also translate the set values for our String columns to English for our investigation. 

In [12]:
autos["vehicle_type"].value_counts().index

Index(['limousine', 'kleinwagen', 'kombi', 'bus', 'cabrio', 'coupe', 'suv',
       'andere'],
      dtype='object')

In [13]:
autos["gearbox"].value_counts().index

Index(['manuell', 'automatik'], dtype='object')

In [14]:
autos["fuel_type"].value_counts().index

Index(['benzin', 'diesel', 'lpg', 'cng', 'hybrid', 'andere', 'elektro'], dtype='object')

In [15]:
autos["unrepaired_damage"].value_counts().index

Index(['nein', 'ja'], dtype='object')

In [16]:
autos["model"].value_counts().index

Index(['golf', 'andere', '3er', 'polo', 'corsa', 'astra', 'passat', 'a4',
       'c_klasse', '5er',
       ...
       'kappa', 'samara', 'elefantino', 'i3', 'kalina', 'serie_2',
       'rangerover', 'serie_3', 'serie_1', 'discovery_sport'],
      dtype='object', length=251)

There are too many unique values for the `model` column so we will not translate every value for our investigation. The columns `vehicle_type`, `gearbox`, `fuel_type`, and `unrepaired_damage` can be translated by mapping their values to their English counterparts. The following translations are sourced from Google Translate.

In [17]:
autos["vehicle_type"] = autos["vehicle_type"].replace(["limousine", "kleinwagen", "kombi", "cabrio", "andere"], ["sedan", "small car", "station wagon", "convertible", "other"])
autos["gearbox"] = autos["gearbox"].replace(["manuell", "automatik"], ["manual", "automatic"])
autos["fuel_type"] = autos["fuel_type"].replace(["benzin", "elektro", "andere"], ["gasoline", "electric", "other"])
autos["unrepaired_damage"] = autos["unrepaired_damage"].replace(["nein", "ja"], ["no", "yes"])
autos["model"] = autos["model"].replace("andere", "other")


In [18]:
autos[["vehicle_type", "gearbox", "fuel_type", "unrepaired_damage", "model"]].head()

vehicle_type    gearbox fuel_type unrepaired_damage  model
0          NaN     manual  gasoline               NaN   golf
1        coupe     manual    diesel               yes    NaN
2          suv  automatic    diesel               NaN  grand
3    small car     manual  gasoline                no   golf
4    small car     manual    diesel                no  fabia

#### Registration Year ####

Another column of interest is the `registration_year` of the vehicle. We can see that the min value of this is `1000` with a max value of `9999`. This data is supposed to represent the year in which the vehicle was first registered. 

In [19]:
autos["registration_year"].describe()

count    371528.000000
mean       2004.577997
std          92.866598
min        1000.000000
25%        1999.000000
50%        2003.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

In [20]:
autos["registration_year"].value_counts().sort_index(ascending=False)

9999    27
9996     1
9450     1
9229     1
9000     5
        ..
1200     1
1111     4
1039     1
1001     1
1000    38
Name: registration_year, Length: 155, dtype: int64

The [Benz Patent Motor Car](https://group.mercedes-benz.com/company/tradition/company-history/1885-1886.html) is widely regarded as the first automobile and was created in 1886. Additionally, the data was scraped in 2016. The discrepancies in these values could be due to issues with the scraper, inaccurate data posted online, or a variety of other issues. Regardless, so we should not consider data from vehicles with a `registration_year` outside of the timeframe of 1886-2016 for accuracy purposes. We can see how many datapoints fall outside of this timeframe to clean up our dataset:

In [21]:
autos.loc[(autos["registration_year"] < 1886) | (autos["registration_year"] > 2016)].shape[0]

14748

In [22]:
(autos.loc[(autos["registration_year"] < 1886) | (autos["registration_year"] > 2016)].shape[0])/len(autos)

0.03969552765874981

Out of 371,528 rows, 14,748 rows or roughly only 4% of our data have `registration_year` values that fall outside of our analysis timeframe. We can safely proceed with removing these datapoints from our dataset.

In [23]:
autos = autos[(autos["registration_year"] >= 1886) & (autos["registration_year"] <= 2016)]

We can verify the data has been cleansed by investigating the new values:

In [24]:
# this code is just to toggle the number of rows displayed for the next code fragment
max_rows = pandas.get_option("display.max_columns")
pandas.set_option("display.max_rows", None)

In [25]:
autos["registration_year"].value_counts(normalize=True).sort_index(ascending=False)

2016    0.027633
2015    0.008358
2014    0.013459
2013    0.017257
2012    0.026397
2011    0.033825
2010    0.034626
2009    0.043744
2008    0.045336
2007    0.049535
2006    0.056702
2005    0.062548
2004    0.055345
2003    0.055701
2002    0.053784
2001    0.056668
2000    0.068813
1999    0.063812
1998    0.050314
1997    0.041219
1996    0.030512
1995    0.027070
1994    0.013655
1993    0.009723
1992    0.008543
1991    0.007876
1990    0.007994
1989    0.003761
1988    0.002705
1987    0.001954
1986    0.001614
1985    0.002206
1984    0.001183
1983    0.001275
1982    0.000939
1981    0.000796
1980    0.001763
1979    0.000874
1978    0.000860
1977    0.000541
1976    0.000541
1975    0.000426
1974    0.000563
1973    0.000575
1972    0.000695
1971    0.000558
1970    0.001037
1969    0.000432
1968    0.000412
1967    0.000398
1966    0.000448
1965    0.000339
1964    0.000224
1963    0.000216
1962    0.000149
1961    0.000143
1960    0.000600
1959    0.000126
1958    0.0000

In [26]:
# resetting the row display
pandas.set_option("display.max_rows", max_rows)

From the remaining data, we can see that a majority of the cars on the market are registered after 1994, which spikes around 2005 before dropping. There are a small number of cars that date back to 1910. This spread is justified as we expect there to be less new cars on the used car market, and cars from decades ago are probably out of style and hard to maintain after so many years.

#### Price ####

From the `price` column description above, we also notice there is quite a large range in terms of price values of the vehicles. Invesgiating this series at this point in our cleansing, we can determine what outliers to exclude from our analysis.

In [27]:
autos["price"].unique().shape

(5552,)

In [28]:
autos["price"].describe()

count    3.567800e+05
mean     1.742357e+04
std      3.657289e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.999000e+03
75%      7.400000e+03
max      2.147484e+09
Name: price, dtype: float64

In [29]:
autos["price"].value_counts().sort_index(ascending=True)

0             10014
1              1119
2                11
3                 7
4                 1
              ...  
32545461          1
74185296          1
99000000          1
99999999         14
2147483647        1
Name: price, Length: 5552, dtype: int64

We have 5,552 unique price values. The average price of a vehicle is \\$17,423 with a minimum value of \\$0 and a max value of \\$2,147,483,647. These values are quite far off from the mean value and can skew the data. 

There are currently 10,014 ads where the price is listed at zero. eBay requires listings to start at \\$1 but bidding can start at \\$0.01. Because the listings with lower end price tags probably have not received bids yet or started auction, we can exclude them from our dataset. Even the cheapest used car should be at least \\$100, so let's take a look at the listings under \\$100.

In [30]:
print(len(autos[autos["price"] < 100]))
autos[autos["price"] < 100].head(15)

12454


date_crawled                                               name  \
7    2016-03-21 18:54:38                       VW_Derby_Bj_80__Scheunenfund   
40   2016-03-26 22:06:17                  Suche_Opel_corsa_a_zu_verschenken   
60   2016-03-29 15:48:15  TAUSCHE_BMW_E38_740i_g._SUV_/_GELÄNDEWAGEN_LES...   
91   2016-03-28 09:37:01  MERCEDES_BENZ_W124_250D_83KW_/_113PS___SCHLACH...   
119  2016-03-20 18:53:27                     Polo_6n_Karosse_zu_verschenken   
157  2016-03-11 18:55:53             Opel_meriva_1.6_16_v_lpg__z16xe_no_OPC   
159  2016-03-22 16:37:59             Fiat_Scudo_Bulli_Werkstattwagen_Diesel   
237  2016-03-16 22:06:24                          Mondeo_Tddi_Teile_gesucht   
279  2016-03-12 07:26:49      Tausche_Youngtimer_Audi_80_mit_knapp_62000_km   
346  2016-03-06 10:06:20                Sehr_sauber_vw_polo_Mit_klimaanlage   
422  2016-03-29 18:51:58          Schlachte_Golf_4._89.000_KM._Blau_1.4_16V   
430  2016-03-11 10:52:49          BMW_323i._e_21_Gesucht__In_jeden_Zustand_   
484  2016-03-24 10:58:40                            Tausche_BMW_top_Zustand   
490  2016-04-04 02:37:41                    BMW_E30_316_318_320_325_GESUCHT   
498  2016-03-24 17:49:19                           Unfallwagen/Totalschaden   

     price   vehicle_type  registration_year    gearbox  power_p_s     model  \
7        0          sedan               1980     manual         50     other   
40       0            NaN               1990        NaN          0     corsa   
60       1            suv               1994     manual        286       NaN   
91       1          sedan               1995     manual        113  e_klasse   
119      0      small car               1999        NaN          0       NaN   
157      0            bus               2004     manual        101    meriva   
159      0            NaN               2006        NaN          0     other   
237      0  station wagon               2001     manual        115    mondeo   
279      1          sedan               1990     manual         90        80   
346      0            NaN               2005        NaN          0      polo   
422      1            NaN               2000        NaN         75      golf   
430      0            NaN               1985        NaN          0       3er   
484      0            NaN               2016  automatic        197       3er   
490      0            NaN               1990        NaN          0       3er   
498      0            NaN               2000     manual          0       NaN   

     odometer_km  registration_month fuel_type           brand  \
7          40000                   7  gasoline      volkswagen   
40        150000                   1  gasoline            opel   
60        150000                  11       NaN  sonstige_autos   
91        150000                   4    diesel   mercedes_benz   
119         5000                   0  gasoline      volkswagen   
157       150000                  10       lpg            opel   
159         5000                   0       NaN            fiat   
237       150000                   0       NaN            ford   
279        70000                   6  gasoline            audi   
346       150000                   0  gasoline      volkswagen   
422        90000                   3  gasoline      volkswagen   
430       150000                   0       NaN             bmw   
484       150000                  12    diesel             bmw   
490         5000                   0       NaN             bmw   
498       150000                   0       NaN            audi   

    unrepaired_damage           ad_created  postal_code            last_seen  
7                  no  2016-03-21 00:00:00        19348  2016-03-25 16:47:58  
40                NaN  2016-03-26 00:00:00        56412  2016-03-27 17:43:34  
60                NaN  2016-03-29 00:00:00        53721  2016-04-06 01:44:38  
91                 no  2016-03-28 00:00:00        40589  2016-04-06 12:15:54  
119           

We see a large amount of $1 listings, which seems pretty unrealistic. 

Meanwhile, the [most expensive car in the world](https://www.driveway.com/learn/buying/what-is-the-most-expensive-car-in-the-world) is documented at \\$28 million dollars. From an article in 2010, eBay listed out the [most expensive cars sold on its platform](https://www.ebay.com/motors/blog/most-expensive-cars-sold-ebay/), with the most expensive vehicle sold having a price tag of \\$3.26 million.

Listing on eBay is tricky, as users are able to list their car under an exorbitantly high value. Selling these is a different story. From eBay community, we can get an idea of one reason as to why this may occur:
> Sometimes a seller wants to reserve an item for a particular buyer, so lists at a very high price no one will pay. Then the buyer it was reserved for makes the agreed on offer and seller accepts it. Actual best offer price it sold for is not revealed. Source:[ebay community](https://community.ebay.com/t5/Ask-a-Mentor/unusual-very-high-price-sold-items-then-same-item-numerous-times/td-p/32580699#:~:text=Sometimes%20a%20seller%20wants%20to,sold%20for%20is%20not%20revealed.) 

Investigating this further, let's take a look at cars marked at over $10 million dollars:

In [31]:
autos[autos["price"] > 10000000].head(15)

date_crawled                                             name  \
14663   2016-04-03 06:03:56                       Tausch_moeglich_Polo_vw_n9   
20746   2016-03-08 15:54:52                               Golf3_Schlachtfest   
54507   2016-03-09 11:40:15                 Tausch_DKW_F8_gegen_Lanz_Bulldog   
56973   2016-03-18 18:52:27  tausche_ford_mondeo_mk3_ghia_gegen_anderes_auto   
69747   2016-03-20 10:59:32                 Suche_Mercedes_Benz_r_klasse_cdi   
77520   2016-03-22 14:58:27                      Tausch_gegen_gleichwertiges   
87799   2016-03-08 20:39:05                                Leasinguebernahme   
120919  2016-03-25 12:43:02              ???_?_???____???_?_???____???_?_???   
127769  2016-03-28 12:51:51                 Golf_4_laesst_sich_nicht_oeffnen   
127820  2016-03-14 17:49:54                                 audi_a6_c5_avant   
129969  2016-03-11 08:37:57                          Kaufe_AUTOS_Zum_basteln   
139052  2016-03-14 12:49:48              Mercedes_Benz_CLS_350_CDI_7G_TRONIC   
150909  2016-04-02 22:41:54                   Suche_Autos_Model_egal_welches   
160530  2016-03-29 12:58:27                                   Golf_4_schrott   
177744  2016-03-31 16:37:00                   Suche_Trabant_auf_Ratenzahlung   

             price   vehicle_type  registration_year    gearbox  power_p_s  \
14663     11111111          coupe               2003     manual         64   
20746     12345678            NaN               1995        NaN          0   
54507     11111111          other               1953     manual         20   
56973     99999999          sedan               2002     manual        131   
69747     99999999            bus               2006        NaN          0   
77520     99999999          sedan               1999  automatic        224   
87799     27322222          sedan               2014     manual        163   
120919    74185296            NaN               2000        NaN          0   
127769    11111111          sedan               2003     manual         75   
127820    99999999  station wagon               1999     manual        165   
129969  2147483647          other               1996     manual        888   
139052    99999999          coupe               2009  automatic        224   
150909    99999999            NaN               1990        NaN          0   
160530    99999999      small car               2001        NaN          0   
177744    11111111            NaN               1960        NaN          0   

           model  odometer_km  registration_month fuel_type           brand  \
14663       polo       150000                   2  gasoline      volkswagen   
20746        NaN       150000                   0       NaN      volkswagen   
54507        NaN       150000                   5       NaN            audi   
56973     mondeo       150000                   1       NaN            ford   
69747      other       150000                   0    diesel   mercedes_benz   
77520   s_klasse       150000                   9  gasoline   mercedes_benz   
87799         c4        40000                   2    diesel         citroen   
120919       145        20000                   0       NaN      alfa_romeo   
127769      golf       125000                   9  gasoline      volkswagen   
127820        a6         5000                  12  gasoline            audi   
129969       NaN       150000                   1     other  sonstige_autos   
139052     other       150000                  11    diesel   mercedes_benz   
150909       NaN         5000                   1       NaN  sonstige_autos   
160530      golf        20000                   0  gasoline      volkswagen   
177744       NaN       150000                   0       NaN         trabant   

       unrepaired_damage           ad_created  postal_code  \
14663                NaN  2016-04-03 00:00:00        46535   
20746                NaN  2016-03-08 00:00:00        18516   
54507                 no  2016-03-

The average price of an Audi A6 C5 (row `127820`) was \\$16,049 from the years 1998-2005 when the car was produced. Source: [classic.com](https://www.classic.com/m/audi/a6/c5/?chart=sales). The listing is of a 1999 model of this vehicle for $99,999,999. This could be an issue with scraping or the user may have listed the car at an extremely high price point which does not reflect market value and has low potential to actually sell. 

Additionally, many of these ads were created recently and delisted before the data scraping completed, as seen in the `ad_created` and `last_seen` columns. Users could be gauging interest on their vehicles without intentions to sell them yet and thus listed at an extremely high price to ensure no one would buy. We can safely discard these datapoints as they do not accurately reflect actual market prices.

We want to really analyze the market for fair market value, so we can utilize the statistical method of calculating interquartile ranges to determine outliers.

`IQR is defined as P75-P25. Strong outliers are values that exist 3 IQRs away from the P25 and P75 values.`

However, our data is skewed by the large number of low listings, so let's drop these first.

In [32]:
autos = autos[autos["price"] > 100]
autos["price"].describe()

count    3.433110e+05
mean     1.810664e+04
std      3.728340e+06
min      1.010000e+02
25%      1.300000e+03
50%      3.200000e+03
75%      7.600000e+03
max      2.147484e+09
Name: price, dtype: float64

In [33]:
iqr = 7600 - 1300
strong_outlier_low = 1300 - 3 * iqr
strong_outlier_high = 7600 + 3 * iqr
print("IQR:", iqr)
print("Strong outliers:", strong_outlier_low, strong_outlier_high)

IQR: 6300
Strong outliers: -17600 26500


Since we have a negative value for the lower end of our outliers, we can just keep the lower threshold to be values greater than $100. Checking how many datapoints fall outside of the higher threshold, we see:

In [34]:
len(autos[autos["price"] > strong_outlier_high])

8763

In [35]:
autos[autos["price"] > strong_outlier_high].head(20)

date_crawled                                               name  \
56   2016-03-21 01:59:07                               BMW_435i_Sport_coupe   
148  2016-03-31 00:59:04  Mercedes_Benz_GLK_250_BlueTEC_4Matic_Standhzg_...   
272  2016-03-29 22:50:53  Volkswagen_Caravelle_Lang_DSG_Navi_Standheiz._...   
275  2016-04-04 23:48:59        Audi_A6_3.0_TDI_competition_S_Line_LED_BOSE   
277  2016-03-11 13:52:52         Mercedes_Benz_CLS_350_CDI_4Matic_7G_TRONIC   
281  2016-03-22 19:56:01  Mercedes_Benz_ML_350_CDI_4Matic_7G_TRONIC_DPF_...   
376  2016-03-11 21:49:52                                       Corvette_Z06   
426  2016-03-18 12:48:27                       Corvette_C1_1959_top_Zustand   
466  2016-03-30 07:54:22                                            Audi_A3   
475  2016-03-20 20:42:07    BMW_X6_xDrive30d_M_Sportpaket_Garantie_bis_2018   
522  2016-03-23 15:42:44                     Audi_A7_3.0_TDI_multitronic_V6   
546  2016-03-10 15:55:03                               Mercedes_Benz_SL_280   
551  2016-03-19 19:46:25  Audi_S5_Exclusive___TipTronic_4.2L_V8_Quattro_...   
556  2016-03-22 12:52:55                   Audi_Q5_3.0_TDI_quattro_S_tronic   
560  2016-03-08 16:53:28       Audi_A4_allroad_quattro_3.0_TDI_DPF_S_tronic   
640  2016-03-06 12:38:01                             Mercedes_Benz_C_63_AMG   
754  2016-04-02 22:57:47        VW_T5_Multivan_Special__AHK__Fahrradtraeger   
768  2016-03-26 12:47:39  Volkswagen_Multivan_4MOTION_MwSt._ausweisbar_S...   
794  2016-03-27 20:38:01                  BMW_435d_Cabrio_xDrive_Sport_Aut.   
844  2016-04-02 09:51:15              Ford_Escort_MK_1_Cosworth_H_Zulassung   

     price   vehicle_type  registration_year    gearbox  power_p_s  \
56   39600          coupe               2014  automatic        306   
148  41900            suv               2014  automatic        204   
272  26999            bus               2013  automatic        140   
275  56900          sedan               2015  automatic        326   
277  48000          coupe               2014  automatic        265   
281  27500            suv               2009  automatic        224   
376  49500    convertible               2013        NaN          0   
426  90500    convertible               1959  automatic        295   
466  29980          sedan               2013     manual        150   
475  43900          sedan               2013  automatic        245   
522  27500          coupe               2010  automatic        204   
546  36500    convertible               1981     manual        185   
551  30900          coupe               2011  automatic        354   
556  28900          sedan               2010  automatic        239   
560  28499  station wagon               2011  automatic        239   
640  41900          sedan               2012  automatic        457   
754  28300            bus               2012     manual        114   
768  37990            bus               2012     manual          0   
794  47999    convertible               2014  automatic        313   
844  34900          sedan               1968     manual        220   

           model  odometer_km  registration_month fuel_type           brand  \
56         other        30000                   7  gasoline             bmw   
148          glk        40000                   9    diesel   mercedes_benz   
272  transporter        80000                  10    diesel      volkswagen   
275           a6         5000                   7    diesel            audi   
277        other        80000                  10    diesel   mercedes_benz   
281     m_klasse       150000                  12    diesel   mercedes_benz   
376          NaN        10000                   6  gasoline  sonstige_autos   
426        other        90000                   7  gasoline       chevrolet   
466           a3        50000                   9    diesel            audi   
475      x_reihe        40000                   4    diesel             bmw   
522        o

We still see a large subset of data, primarily newer cars, that could be viable datapoints with these calculations, so let's refine the outlier multiplier a bit. 

In [36]:
strong_outlier_high = 7.500000e+03 + 3 * 10 * iqr
print("Strong outlier high:", strong_outlier_high)

Strong outlier high: 196500.0


In [37]:
len(autos[autos["price"] > strong_outlier_high])

167

In [38]:
autos[autos["price"] > strong_outlier_high].head(20)

date_crawled                                               name  \
1846   2016-03-10 22:47:05  BMW_M1_Museumsfahrzeug_Neuwagenzustand_Glossy_...   
10649  2016-04-05 09:36:18                      Porsche_996_GT2_MK2_Clubsport   
14663  2016-04-03 06:03:56                         Tausch_moeglich_Polo_vw_n9   
16889  2016-03-15 16:45:45                   Ford_Mondeo_Kombi_TÜV_abgelaufen   
20143  2016-03-18 00:55:53                         Porsche_911R_Einer_von_911   
20746  2016-03-08 15:54:52                                 Golf3_Schlachtfest   
20818  2016-03-14 13:56:15  Porsche_991_Carrera_S__50_Jahre_911___X51__430...   
21467  2016-03-19 19:45:02  Porsche_911_911R_1_of_20_ORIGINAL_R_nur_798KG!...   
26327  2016-03-21 19:43:54                                      Porsche_911_R   
28988  2016-03-21 10:50:12                                        Porsche_997   
29866  2016-03-21 11:50:40                                    Porsche_997_GT2   
38445  2016-03-21 19:53:52                                      2_VW_Busse_T3   
41980  2016-03-11 09:37:17                                      Maserati_MC12   
45055  2016-04-02 00:06:21  Ferrari_LaFerrari_Deutsche_Auslieferung._A_Cel...   
45833  2016-03-15 10:53:11                            Porsche_911_Turbo__930_   
46503  2016-03-09 00:55:57                                        Porsche_911   
53507  2016-03-11 18:57:57         Suzuki_SJ_Samurai_413_Van__Kein_Lada_Niva_   
54295  2016-03-08 11:51:40                       Mazda_5_2.0_CD_DPF_Exclusive   
54507  2016-03-09 11:40:15                   Tausch_DKW_F8_gegen_Lanz_Bulldog   
55825  2016-03-29 20:47:24                                   Andere_Horch_830   

          price   vehicle_type  registration_year    gearbox  power_p_s  \
1846     579000          coupe               1980     manual        277   
10649    420000          coupe               2004     manual        483   
14663  11111111          coupe               2003     manual         64   
16889   1000000  station wagon               1998        NaN          0   
20143   1250000          coupe               2016     manual        500   
20746  12345678            NaN               1995        NaN          0   
20818    225000          coupe               2014  automatic        430   
21467   9999999          coupe               1967     manual        215   
26327    600000          coupe               2016     manual        500   
28988    220000          coupe               2008     manual        415   
29866    245000          coupe               2008     manual        530   
38445    999999            bus               1981     manual         70   
41980   2995000          coupe               2005        NaN        630   
45055   2795000          coupe               2016        NaN        963   
45833   1000000          coupe               1981     manual        300   
46503    199900    convertible               1972     manual        207   
53507   1000000            suv               1992     manual         69   
54295  10000000            bus               2007     manual        143   
54507  11111111          other               1953     manual         20   
55825    235000    convertible               1933     manual         69   

             model  odometer_km  registration_month fuel_type           brand  \
1846         other        20000                  12  gasoline             bmw   
10649          911        50000                   4  gasoline         porsche   
14663         polo       150000                   2  gasoline      volkswagen   
16889       mondeo       150000                   0  gasoline            ford   
20143          911         5000                   3  gasoline         porsche   
20746          NaN       150000                   0       NaN      volkswagen   
20818          911        10000                   1  gasoline         porsche   
21467          911        50000                  10  gasoline         porsche   
26327  

There are a number of supercars in this dataset, but overall we can proceed with dropping this data. Supercars are rare and skew the data for our investigation as they are unlikely to have a good price-to-value ratio to the majority of prospective buyers. Out of hundreds of thousands of datapoints, there are only 167 cars that fall above this value. 

In [39]:
autos = autos[(autos["price"] > 0) & (autos["price"] <= strong_outlier_high)]

We can verify that the new prices fall within our dedicated range of \\$1 - \\$195,000:

In [40]:
autos["price"].value_counts().sort_index(ascending=False)

195900     1
195000     3
194000     1
192500     2
190000     1
          ..
111       15
110       35
108        1
105        3
101        4
Name: price, Length: 5401, dtype: int64

In [41]:
autos["price"].describe()

count    343144.000000
mean       6006.602523
std        8352.530174
min         101.000000
25%        1300.000000
50%        3200.000000
75%        7600.000000
max      195900.000000
Name: price, dtype: float64

#### Dates ####

Taking a look at the date columns, we can see that the data is provided in terms of exact time stamps down to the second.

In [42]:
autos[["date_crawled", "ad_created", "last_seen"]][0:5]

date_crawled           ad_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

For the purpose of our analysis, we want to be able to see the dates just in terms of the exact day. From the String object, only the first ten characters of the string correspond to the day, so we can leverage this to examine the date information. We should also convert these columns to `datetime` values.

In [43]:
autos["date_crawled"] = pandas.to_datetime(autos["date_crawled"].str[:10])
autos["ad_created"] = pandas.to_datetime(autos["ad_created"].str[:10])
autos["last_seen"] = pandas.to_datetime(autos["last_seen"].str[:10])
autos[["ad_created", "date_crawled", "last_seen"]].info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 343144 entries, 0 to 371527
Data columns (total 3 columns):
 #   Column        Non-Null Count   Dtype         
---  ------        --------------   -----         
 0   ad_created    343144 non-null  datetime64[ns]
 1   date_crawled  343144 non-null  datetime64[ns]
 2   last_seen     343144 non-null  datetime64[ns]
dtypes: datetime64[ns](3)
memory usage: 10.5 MB


date_crawled                            name  price vehicle_type  \
0   2016-03-24                      Golf_3_1.6    480          NaN   
1   2016-03-24            A5_Sportback_2.7_Tdi  18300        coupe   
2   2016-03-14  Jeep_Grand_Cherokee_"Overland"   9800          suv   
3   2016-03-17              GOLF_4_1_4__3TÜRER   1500    small car   
4   2016-03-31  Skoda_Fabia_1.4_TDI_PD_Classic   3600    small car   

   registration_year    gearbox  power_p_s  model  odometer_km  \
0               1993     manual          0   golf       150000   
1               2011     manual        190    NaN       125000   
2               2004  automatic        163  grand       125000   
3               2001     manual         75   golf       150000   
4               2008     manual         69  fabia        90000   

   registration_month fuel_type       brand unrepaired_damage ad_created  \
0                   0  gasoline  volkswagen               NaN 2016-03-24   
1                   5    diesel        audi               yes 2016-03-24   
2                   8    diesel        jeep               NaN 2016-03-14   
3                   6  gasoline  volkswagen                no 2016-03-17   
4                   7    diesel       skoda                no 2016-03-31   

   postal_code  last_seen  
0        70435 2016-04-07  
1        66954 2016-04-07  
2        90480 2016-04-05  
3        91074 2016-03-17  
4        60437 2016-04-06

### Data Analysis ###

#### Dates ####
Picking up where we left off, we can now investigate the date information and distribution of `date_crawled`, `ad_created`, and `last_seen` columns.

In [44]:
autos["date_crawled"].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.025680
2016-03-06    0.014527
2016-03-07    0.035737
2016-03-08    0.033578
2016-03-09    0.034169
                ...   
2016-04-03    0.038981
2016-04-04    0.037722
2016-04-05    0.012651
2016-04-06    0.003130
2016-04-07    0.001580
Name: date_crawled, Length: 34, dtype: float64

In [45]:
autos["date_crawled"].value_counts(normalize=True, dropna=False).describe()

count    34.000000
mean      0.029412
std       0.009689
min       0.001580
25%       0.030126
50%       0.032724
75%       0.035076
max       0.038981
Name: date_crawled, dtype: float64

The data crawling began on March 5, 2016 and concluded on April 7, 2016, running daily during this time period. The data is pretty uniformly distributed across the days, dropping off more towards the end of the scraping period. On average, around 3% of the data available was scraped on each day.

In [46]:
autos["ad_created"].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
                ...   
2016-04-03    0.039193
2016-04-04    0.037795
2016-04-05    0.011494
2016-04-06    0.003121
2016-04-07    0.001515
Name: ad_created, Length: 113, dtype: float64

In [47]:
autos["ad_created"].value_counts(normalize=True, dropna=False).describe()

count    113.000000
mean       0.008850
std        0.014435
min        0.000003
25%        0.000006
50%        0.000023
75%        0.017121
max        0.039193
Name: ad_created, dtype: float64

The earliest ad scraped in this dataset was created on March 10, 2014 and the latest ad available was created on the final day of scraping, April 7, 2016. Thus, our data consists of ads created throughout early 2014 through early 2016, providing us roughly around 2 years of datapoints. More ads were created recently in terms of the time that the data was scraped from. This makes sense as used cars from older ads are more likely to have been sold already and delisted.

In [48]:
autos["last_seen"].value_counts(normalize=True, dropna=False).sort_index(ascending=True)

2016-03-05    0.001268
2016-03-06    0.004045
2016-03-07    0.005196
2016-03-08    0.007825
2016-03-09    0.009710
                ...   
2016-04-03    0.025167
2016-04-04    0.025377
2016-04-05    0.127585
2016-04-06    0.220767
2016-04-07    0.131580
Name: last_seen, Length: 34, dtype: float64

In [49]:
autos["last_seen"].value_counts(normalize=True, dropna=False).describe()

count    34.000000
mean      0.029412
std       0.043761
min       0.001268
25%       0.012324
50%       0.018921
75%       0.023801
max       0.220767
Name: last_seen, dtype: float64

Ads are removed when the listing has been sold or delisted. Most of the ads were seen relatively frequently by the crawler online at the time of the data collection, with a spike in the last three days of the data crawling. Thus, most of the ads we are investigating were still active on the site at the time of the data collection.

#### Brands ####
Does the brand of a car affect the price and value of the vehicle? We can determine the answer to this question by aggregating the data across the `brand` values in our dataset for investigation:

In [50]:
autos["brand"].value_counts()

volkswagen       72689
bmw              37819
opel             36358
mercedes_benz    33406
audi             30873
                 ...  
daewoo             493
trabant            461
lancia             440
rover              431
lada               206
Name: brand, Length: 40, dtype: int64

In [51]:
brand_mean_price = {}
brands = autos["brand"].value_counts().index
for brand in brands:
    brand_mean_price[brand] = autos[autos["brand"] == brand]["price"].sum()/len(autos[autos["brand"] == brand])
brand_mean_price_series = pandas.Series(brand_mean_price)

In [52]:
brand_mean_price_series.sort_values(ascending=False)[:10]

porsche           39314.920612
land_rover        17070.102530
sonstige_autos    14074.720461
jaguar            13562.944351
jeep              11287.464901
mini              10105.366102
audi               9125.106274
mercedes_benz      8536.637670
bmw                8482.138925
chevrolet          7146.441176
dtype: float64

In [53]:
brand_mean_price_series.sort_values(ascending=True)[:10]

daewoo      1051.152130
rover       1636.406032
daihatsu    1826.499312
trabant     2009.991323
renault     2473.436618
fiat        2933.144517
opel        3014.385692
lada        3206.436893
peugeot     3298.947836
lancia      3370.943182
dtype: float64

In [54]:
brand_mean_price_series.describe()

count       40.000000
mean      6419.505936
std       6398.512921
min       1051.152130
25%       3425.064380
50%       4415.777258
75%       6697.920224
max      39314.920612
dtype: float64

In [55]:
brand_mean_price_series["volkswagen"]

5448.756772001266

From this data, we can see that brands such as Porsche and Land Rover have the highest pricing. Meanwhile, Lancia and Peugot are lowest in terms of pricing. Volkswagen seems to be the most popular brand listed and price average falls in the middle of the range, above the 50th percentile, but below the mean across all car brands.

#### Value ####

We can determine a car's value to be dependent on the pricing of a vehicle and how much usage the vehicle has been through. Leveraging the `odometer_km` column, we can determine how many kilometers have been driven at the time of sale for the vehicle. Let's inspect this column now:

In [56]:
autos["odometer_km"].unique().shape

(13,)

In [57]:
autos["odometer_km"].value_counts().sort_index()

5000        4989
10000       1738
20000       5314
30000       5768
40000       6201
50000       7340
60000       8415
70000       9408
80000      10566
90000      11947
100000     14892
125000     35810
150000    220756
Name: odometer_km, dtype: int64

In [58]:
autos["odometer_km"].describe()

count    343144.000000
mean     125601.802742
std       39624.124257
min        5000.000000
25%      100000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

From the dataset, there are 15 unique values for the `odometer_km` colummn. The data is not precise, but givees us a range of understanding how many kilometers have been driven on the vehicle. The mean kilometers driven is 125,601 and the min is 5,000 with a max of 150,000. There are a large number of vehicles with 150,000 kilometers driven as we see our 50th percentile and 75th percentile also have values of 150,000. 

A car that has been driven more than its counterpart should have a lower price. This is due to wear and tear of the vehicle through usage.

In [59]:
mileage_mean_price = {}
mileage_values = autos["odometer_km"].value_counts().index
for mileage_value in mileage_values:
    mileage_mean_price[mileage_value] = autos[autos["odometer_km"] == mileage_value]["price"].sum()/len(autos[autos["odometer_km"] == mileage_value])
mileage_mean_price_series = pandas.Series(mileage_mean_price)
mileage_mean_price_series.sort_index()

5000       8705.351774
10000     21455.326237
20000     17223.935265
30000     16042.897365
40000     15138.993227
50000     13450.720708
60000     12077.202971
70000     11013.962266
80000      9705.476812
90000      8707.365531
100000     7934.681708
125000     6351.728149
150000     3833.635321
dtype: float64

There does not seem to be a direct correlation between mileage and average price of the vehicle, as cars with only 5,000 kilometers driven have some of the lowest price averages. However, the general trend is that the price of the vehicle declines with the number of kilometers driven on the vehicle. 

We can see if this pattern affects certain brands of cars' pricing:

In [60]:
brand_mean_mileage = {}
brands = autos["brand"].value_counts().index
for brand in brands:
    brand_mean_mileage[brand] = autos[autos["brand"] == brand]["odometer_km"].sum()/len(autos[autos["brand"] == brand])
brand_mean_mileage_series = pandas.Series(brand_mean_mileage)
brand_mean_mileage_series.sort_values(ascending=False)

saab              141623.762376
volvo             138311.626429
rover             135278.422274
chrysler          134246.782740
bmw               133019.646210
                      ...      
mini               93320.493066
sonstige_autos     88802.433557
dacia              86060.254925
lada               74320.388350
trabant            55780.911063
Length: 40, dtype: float64

In [61]:
brand_means_df = pandas.DataFrame(brand_mean_price_series, columns=["mean_price"])
brand_means_df["mean_mileage"] = brand_mean_mileage_series
brand_means_df["price_to_mileage_ratio"] = brand_means_df["mean_price"]/brand_means_df["mean_mileage"]
brand_means_df.sort_values("price_to_mileage_ratio", ascending=False)

mean_price   mean_mileage  price_to_mileage_ratio
porsche         39314.920612  100963.653754                0.389397
sonstige_autos  14074.720461   88802.433557                0.158495
land_rover      17070.102530  119114.513981                0.143308
jaguar          13562.944351  123532.883642                0.109792
mini            10105.366102   93320.493066                0.108287
...                      ...            ...                     ...
opel             3014.385692  128769.321745                0.023409
renault          2473.436618  127891.785296                0.019340
daihatsu         1826.499312  118651.994498                0.015394
rover            1636.406032  135278.422274                0.012097
daewoo           1051.152130  123681.541582                0.008499

[40 rows x 3 columns]

In [62]:
brand_means_df.loc["trabant"]

mean_price                 2009.991323
mean_mileage              55780.911063
price_to_mileage_ratio        0.036034
Name: trabant, dtype: float64

From this ratio, we can see that on average the more expensive cars have aa higher price to mileage ratio, meaning that cars with more kilometers driven are still on average more expensive for the more expensive brands. On average, the brand Trabant has the least miles driven on the vehicle and is on the cheaper end of the pricing spectrum.

Additionally, we should consider damages to the vehicle affecting vehicle value. Car repair damages can vary in cost from below a hundred dollars up to thousands of dollars. Additionally, older and rarer cars can require special parts that can increase the costs.We can use the `unrepaired_damage` column:

In [63]:
autos["unrepaired_damage"].value_counts().sort_index()

no     252126
yes     32255
Name: unrepaired_damage, dtype: int64

In [64]:
brands_damage_status = autos[["brand", "unrepaired_damage"]].value_counts().index
brand_damaged_mean_price = {}
brand_undamaged_mean_price = {}
for brand, damage_status in brands_damage_status:
    if damage_status == "yes":
        brand_damaged_mean_price[brand] = autos[(autos["brand"] == brand) & (autos["unrepaired_damage"] == damage_status)]["price"].sum()/len(autos[(autos["brand"] == brand) & (autos["unrepaired_damage"] == damage_status)])
    else:
        brand_undamaged_mean_price[brand] = autos[(autos["brand"] == brand) & (autos["unrepaired_damage"] == damage_status)]["price"].sum()/len(autos[(autos["brand"] == brand) & (autos["unrepaired_damage"] == damage_status)])
brand_damaged_mean_price_series = pandas.Series(brand_damaged_mean_price)
brand_undamaged_mean_price_series = pandas.Series(brand_undamaged_mean_price)
brand_damaged_status_mean_price_df = pandas.DataFrame(brand_damaged_mean_price_series, columns=["mean_price_damaged"])
brand_damaged_status_mean_price_df["mean_price_undamaged"] = brand_undamaged_mean_price_series
brand_damaged_status_mean_price_df["damage_difference"] = brand_damaged_status_mean_price_df["mean_price_undamaged"] - brand_damaged_status_mean_price_df["mean_price_damaged"]
brand_damaged_status_mean_price_df.sort_values("damage_difference")

mean_price_damaged  mean_price_undamaged  damage_difference
daewoo                  634.806818           1215.860927         581.054109
daihatsu                921.318681           2161.763265        1240.444584
rover                   818.259740           2143.449393        1325.189652
renault                1062.816004           3048.013176        1985.197172
lada                   1732.400000           3811.805755        2079.405755
...                            ...                   ...                ...
jeep                   4230.609375          12466.301724        8235.692349
jaguar                 3930.454545          15966.971554       12036.517008
sonstige_autos         5155.431280          17839.034294       12683.603014
land_rover             5550.392157          19383.724756       13833.332599
porsche               19159.780488          41705.333514       22545.553026

[40 rows x 3 columns]

In [65]:
brand_damaged_status_mean_price_df.loc["trabant"]

mean_price_damaged       784.480000
mean_price_undamaged    2964.387097
damage_difference       2179.907097
Name: trabant, dtype: float64

From this, we can see that unrepaired damages does greatly decrease the value of the car, most of the times over half of the price of the vehicle. Often, the steep discount can indicate the amount the damages cost to fix exceed the total cost of the damaged vehicle. Thus, for our case to analyze what car to buy when purchasing a usable car at a great price, we should not consider damaged vehicles. Trabant still has a great price for the undamaged vehicles. We can confirm this against the other car brands:

In [66]:
brand_mean_price_no_damages = {}
brands = autos["brand"].value_counts().index
for brand in brands:
    brand_mean_price_no_damages[brand] = autos[(autos["brand"] == brand) & (autos["unrepaired_damage"] == "no")]["price"].sum()/len(autos[(autos["brand"] == brand) & (autos["unrepaired_damage"] == "no")])
brand_mean_price_no_damages_series = pandas.Series(brand_mean_price_no_damages)
brand_mean_price_no_damages_series.sort_values(ascending=False)

porsche           41705.333514
land_rover        19383.724756
sonstige_autos    17839.034294
jaguar            15966.971554
jeep              12466.301724
                      ...     
renault            3048.013176
trabant            2964.387097
daihatsu           2161.763265
rover              2143.449393
daewoo             1215.860927
Length: 40, dtype: float64

In [67]:
brand_mean_mileage_no_damages = {}
brands = autos["brand"].value_counts().index
for brand in brands:
    brand_mean_mileage_no_damages[brand] = autos[(autos["brand"] == brand) & (autos["unrepaired_damage"] == "no")]["odometer_km"].sum()/len(autos[(autos["brand"] == brand) & (autos["unrepaired_damage"] == "no")])
brand_mean_mileage_no_damages_series = pandas.Series(brand_mean_mileage_no_damages)
brand_mean_mileage_no_damages_series.sort_values(ascending=False)

saab              141447.028424
volvo             136532.877882
rover             134008.097166
chrysler          132882.483370
bmw               131238.978723
                      ...      
mini               90797.050562
sonstige_autos     89490.169182
dacia              84031.117397
lada               69748.201439
trabant            53963.133641
Length: 40, dtype: float64

## Conclusion ##
On the German car market, the best brand to purchase from in terms of value and lowest cost of a drivable vehicle with no damages required to be fixed, Trabant has the best price for the amount of miles the car has been driven. We can see our options here, sorted by lowest to highest price.

In [68]:
autos[(autos["brand"] == brand) & (autos["unrepaired_damage"] == "no")].sort_values("price")

date_crawled                                               name  price  \
205116   2016-04-01                              Auto_Lada_Kalina_1118    500   
49477    2016-03-29  Lada_111_Kombi_1_5_Li_mit_Anhaengerkupplung_TÜ...    550   
205733   2016-03-08  Lada_111_Kombi_1_5_GTE_16V_TÜV_1/2017_Top_gepf...    600   
277907   2016-03-15                                 Lada_Lada_2111_16V    600   
51289    2016-03-10                        Lada_2111_16V_"GTe"_KLIMA_"    700   
...             ...                                                ...    ...   
355020   2016-03-10       Lada_Niva_4x4_Deutsches_Fzg_mit_AHK_1900_KG!   9450   
327967   2016-03-30    Lada_OLDTIMER_LADA_NOVA_VAZ_2105_Variante_1300S   9900   
336371   2016-03-14                                          Lada_Niva   9999   
130467   2016-04-06  Lada_Niva_/_Taiga_4x4_im_Safari_Dekor_mit_1_9T...  12000   
187368   2016-04-04  Lada_Niva_Urban__Dotz__AHK__Lederlenkrad__Schn...  12390   

         vehicle_type  registration_year gearbox  power_p_s   model  \
205116          sedan               2007  manual          0  kalina   
49477   station wagon               2002  manual         77   other   
205733  station wagon               2002  manual         91   other   
277907  station wagon               2005  manual         90   other   
51289   station wagon               2004  manual         91   other   
...               ...                ...     ...        ...     ...   
355020            suv               2014  manual         83    niva   
327967          sedan               1985  manual         65   other   
336371            suv               2014  manual         83    niva   
130467            suv               2014  manual         83    niva   
187368            suv               2016  manual         83    niva   

        odometer_km  registration_month fuel_type brand unrepaired_damage  \
205116       100000                   8  gasoline  lada                no   
49477         60000                   8  gasoline  lada                no   
205733       100000                   1  gasoline  lada                no   
277907        80000                   6  gasoline  lada                no   
51289         70000                   7  gasoline  lada                no   
...             ...                 ...       ...   ...               ...   
355020        20000                   4  gasoline  lada                no   
327967         5000                   7  gasoline  lada                no   
336371         5000                   7  gasoline  lada                no   
130467        20000                  10  gasoline  lada                no   
187368        10000                   1  gasoline  lada                no   

       ad_created  postal_code  last_seen  
205116 2016-04-01        59192 2016-04-03  
49477  2016-03-29        93104 2016-04-05  
205733 2016-03-08        73733 2016-03-12  
277907 2016-03-15        67661 2016-04-05  
51289  2016-03-10        13509 2016-03-12  
...           ...          ...        ...  
355020 2016-03-10        90443 2016-03-12  
327967 2016-03-30        12277 2016-04-07  
336371 2016-03-14        71384 2016-04-06  
130467 2016-04-06        72202 2016-04-06  
187368 2016-04-04        44575 2016-04-06  

[139 rows x 16 columns]